src/v/kafka/server/group.cc

```diff
@@ -3336,4 +3339,154 @@ void group::update_subscriptions() {
+             * <kafka>
+             * the group is stable and consumers exist.
+             *
+             * if an offset's topic is subscribed to and retention period has
+             * passed since the last commit timestamp, then expire the offset.
+             * offsets with pending commit are not expired.
+             * </kafka>
+             */
+            return filter_expired_offsets(
+              retention_period,
+              [this](const auto& topic) {
+                  return _subscriptions.value().contains(topic);
+              },
+              [](const offset_metadata& md) { return md.commit_timestamp; });
+
+        } else if (in_state(group_state::empty)) {
+            /*
+             * policy description from kafka source:
+             *
+             * <kafka>
+             * the group contains no consumers.
+             *
+             * if current state timestamp exists and retention period has passed
+             * since group became empty, expire all offsets with no pending
+             * offset commit.
+             *
+             * if there is no current state timestamp (old group metadata
+             * schema) and retention period has passed since the last commit
+             * timestamp, expire the offset
+             * </kafka>
+             */
+            return filter_expired_offsets(
+              retention_period,
+              not_subscribed,
+              [this](const offset_metadata& md) {
+                  if (_state_timestamp.has_value()) {
+                      return _state_timestamp.value();
+                  } else {
+                      return md.commit_timestamp;
+                  }
+              });
+        } else {
+            return {};
+        }
+    } else {
+        /*
+         * policy description from kafka source:
+         *
+         * <kafka>
+         * there is no configured protocol type, so this is a standalone/simple
+         * consumer that Kafka uses for offset storage only.
+         *
+         * expire offsets with no pending offset commits for which the retention
+         * period has passed since their last commit.
+         * </kafka>
+         */
+        return filter_expired_offsets(
+          retention_period, not_subscribed, [](const offset_metadata& md) {
+              return md.commit_timestamp;
+          });
+    }
+}
+
+std::vector<model::topic_partition>
+group::delete_expired_offsets(std::chrono::seconds retention_period) {
+    /*
+     * collect and delete expired offsets
+     */
+    auto offsets = get_expired_offsets(retention_period);
+    for (const auto& offset : offsets) {
+        vlog(_ctxlog.debug, "Expiring group offset {}", offset);
+        _offsets.erase(offset);
+    }
+
+    /*
+     * maybe mark the group as dead
+     */
+    const auto has_offsets = [this] {
+        return !_offsets.empty() || !_pending_offset_commits.empty()
+               || !_volatile_txs.empty() || !_prepared_txs.empty();
```

Comment:
instead of `!_prepared_txs.empty()` we should use `!_tx_seqs.empty()`

Reply:
@rystsov and keep volatile_txs check?

Reply:
Fixed!